### PR TITLE
Disable cursor opacity animation on macOS, make iOS cursor animation discrete

### LINKED ISSUE #85781: Too much CPU usage with TextField caret on macOS
## Details

**Target Platform: macOS**
**Target OS version/browser: Big sur 11.4 (20F71)**
**Devices: MacBook Pro (16-inch, 2019)**

A simple flutter application (which only contains a `TextField` widget), but it takes up **10%** of the CPU.

image

## Reproduce

1. copy this code to your example project `main.dart`.
```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const Scaffold(
        body: TextField(autofocus: true),
      ),
    );
  }
}

```

2. `flutter run -d macos --profile`

3.  We could use command top to check our flutter_app CPU usage: ```top -pid `pgrep flutter_app` ```

## Logs

<details>
<summary>Logs</summary>

```
➜  flutter_app flutter analyze
Analyzing flutter_app...                                                
No issues found! (ran in 1.3s)
```

```
➜  flutter_app flutter doctor -v
[✓] Flutter (Channel master, 2.3.0-17.0.pre.632, on macOS 11.4 20F71 darwin-x64, locale zh-Hans-CN)
    • Flutter version 2.3.0-17.0.pre.632 at /Users/yangbin/dev/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision dbd3f0e0d8 (5 hours ago), 2021-07-02 06:06:03 -0400
    • Engine revision 314cf7a4f2
    • Dart version 2.14.0 (build 2.14.0-269.0.dev)

[!] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for more details.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.5.1, Build version 12E507
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Users/yangbin/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/202.7351085/Android
      Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] Android Studio (version 4.2)
    • Android Studio at /Users/yangbin/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/202.7351085/Android
      Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] Android Studio (version 4.2)
    • Android Studio at /Users/yangbin/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/202.7322048/Android
      Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] IntelliJ IDEA Ultimate Edition (version 2021.1.2)
    • IntelliJ at /Users/yangbin/Applications/JetBrains Toolbox/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 57.0.5
    • Dart plugin version 211.7233

[✓] IntelliJ IDEA Ultimate Edition (version 2021.1.2)
    • IntelliJ at /Users/yangbin/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/211.7442.40/IntelliJ IDEA.app
    • Flutter plugin version 56.0.5
    • Dart plugin version 211.7233

[✓] IntelliJ IDEA Ultimate Edition (version 2021.1.1)
    • IntelliJ at /Users/yangbin/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/211.7142.45/IntelliJ IDEA.app
    • Flutter plugin version 56.0.5
    • Dart plugin version 211.7233

[✓] VS Code (version 1.57.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.23.0

[!] Proxy Configuration
    • HTTP_PROXY is set
    ! NO_PROXY is not set

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 11.4 20F71 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 91.0.4472.114

! Doctor found issues in 2 categories.
```

</details>

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
- Disable animated blinking cursor on macOS. The caret is an `NSView` that gets removed/added back periodically in AppKit text input views, its opacity doesn't change. This fixes https://github.com/flutter/flutter/issues/85781, CPU usage stays below/around 1% in my test app.
- Make the iOS cursor blinking opacity animation discrete to match the native. This should also make the animation more energy efficient as we're no longer interpolating between the keyframes.
- Disable "briefly show password" on iOS, since secured input fields don't seem to be doing that on iOS 15.4 

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 1]
```diff
@@ -3489,7 +3543,7 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
       text = widget.obscuringCharacter * text.length;
       // Reveal the latest character in an obscured field only on mobile.
```
Comment (cyanglaz) [inline packages/flutter/lib/src/widgets/editable_text.dart:3544]: This comment has to be updated. Also a good place to explain why iOS is excluded, maybe add the issue link?

File: packages/flutter/test/widgets/editable_text_cursor_test.dart [hunk 2]
```diff
@@ -166,61 +166,56 @@ void main() {
...
+    await verifyKeyFrame(opacity: 0.75, at: 537500);
+    await verifyKeyFrame(opacity: 0.5,  at: 575000);
+    await verifyKeyFrame(opacity: 0.25, at: 612500);
+    await verifyKeyFrame(opacity: 0.0,  at: 650000);
+    await verifyKeyFrame(opacity: 0.0,  at: 850000);
+    await verifyKeyFrame(opacity: 0.25, at: 887500);
+    await verifyKeyFrame(opacity: 0.5,  at: 925000);
+    await verifyKeyFrame(opacity: 0.75, at: 962500);
+    await verifyKeyFrame(opacity: 1.0,  at: 1000000);
+  }, variant: const TargetPlatformVariant(<TargetPlatform>{ TargetPlatform.iOS }));
```
Comment (cyanglaz) [inline packages/flutter/test/widgets/editable_text_cursor_test.dart:218]: Didn't we need to keep the macOS version of the test?

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 3]
```diff
@@ -1597,7 +1668,14 @@ class EditableText extends StatefulWidget {
...
 class EditableTextState extends State<EditableText> with AutomaticKeepAliveClientMixin<EditableText>, WidgetsBindingObserver, TickerProviderStateMixin<EditableText>, TextSelectionDelegate implements TextInputClient, AutofillClient {
   Timer? _cursorTimer;
-  bool _targetCursorVisibility = false;
+  AnimationController get _cursorBlinkOpacityController {
+    return _backingCursorBlinkOpacityController ??= AnimationController(
+      vsync: this,
+    )..addListener(_onCursorColorTick);
+  }
+  AnimationController? _backingCursorBlinkOpacityController;
+  late final Simulation _iosBlinkCursorSimulation = _DiscreteKeyFrameSimulation.iOSBlinkingCaret();
```
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/editable_text.dart:1677]: The simulation will only be initialized when `widget.cursorOpacityAnimates` is true?

**Review by LongCatIsLooong — COMMENTED** [review]
Changed the simulation to no longer wrap around when the input `time` becomes >= maxDuration. `EditableTextState` ends the animation anyways.

**Review by cyanglaz — APPROVED** [review]
LGTM

File: packages/flutter/test/widgets/editable_text_cursor_test.dart [hunk 4]
```diff
@@ -956,4 +970,40 @@ void main() {
     );
     EditableText.debugDeterministicCursor = false;
   }, variant: const TargetPlatformVariant(<TargetPlatform>{ TargetPlatform.iOS,  TargetPlatform.macOS }));
+
+  testWidgets('password briefly does not show last character on Android if turned off', (WidgetTester tester) async {
```
Comment (justinmc) [inline packages/flutter/test/widgets/editable_text_cursor_test.dart:974]: This mentions Android but it's run on the current platform?  Shouldn't it fail on iOS?

**Review by justinmc — APPROVED** [review]
LGTM :+1:

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 5]
```diff
@@ -301,6 +297,91 @@ class ToolbarOptions {
   final bool selectAll;
 }
 
+// A time-value pair that represents a key frame in an animation.
+class _KeyFrame {
+  const _KeyFrame(this.time, this.value);
+  // Values extracted from iOS 15.4 UIKit.
```
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/editable_text.dart:303]: By inspecting its `CALayer`'s properties. The `CAAnimation`s are stored in an array.

File: packages/flutter/lib/src/widgets/editable_text.dart (hunk 1 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/editable_text.dart:3554]: Yeah that was for Android only.

File: packages/flutter/test/widgets/editable_text_cursor_test.dart (hunk 4 above)
Comment (LongCatIsLooong) [reply packages/flutter/test/widgets/editable_text_cursor_test.dart:974]: Renamed the test

File: packages/flutter/test/widgets/editable_text_cursor_test.dart [hunk 6]
```diff
@@ -166,61 +166,75 @@ void main() {
...
-    await tester.pump(const Duration(milliseconds: 50));
-    // Cursor starts coming back.
-    expect(renderEditable.cursorColor!.alpha, 79);
-    expect(renderEditable, paints..rrect(color: const Color(0x4f2196f3)));
-  }, variant: const TargetPlatformVariant(<TargetPlatform>{ TargetPlatform.iOS,  TargetPlatform.macOS }));
+    for (int i = 0; i < 40; i += 1) {
+      await tester.pump(const Duration(milliseconds: 100));
+      expect(tester.hasRunningAnimations, false);
+    }
+  }, variant: TargetPlatformVariant(TargetPlatform.values.toSet()..remove(TargetPlatform.iOS)));
```
[the remark this answers is not in the document]
Comment (antholeole) [reply packages/flutter/test/widgets/editable_text_cursor_test.dart:237]: Almost exactly how 106216 is implemented! Depending on what gets in first, I’ll either change this to use the new API or this should use the new API. 

Another good use case :).

**Pull request merged**